ownCloud/Tasks/InstantMediaUploadTaskExtension.swift

```diff
@@ -0,0 +1,203 @@
+	override class var locations : [OCExtensionLocationIdentifier]? { return [.appDidComeToForeground] }
+	override class var features : [String : Any]? { return [ FeatureKeys.photoLibraryChanged : true, FeatureKeys.runOnWifi : true] }
+
+	private var uploadDirectoryTracking: OCCoreItemTracking?
+
+	override func run(background:Bool) {
+		guard let userDefaults = OCAppIdentity.shared.userDefaults else { return }
+
+		guard userDefaults.instantUploadPhotos == true || userDefaults.instantUploadVideos == true else { return }
+
+		guard let bookmarkUUID = userDefaults.instantUploadBookmarkUUID else { return }
+
+		guard let path = userDefaults.instantUploadPath else { return }
+
+		if let bookmark = OCBookmarkManager.shared.bookmark(for: bookmarkUUID) {
+
+			OCCoreManager.shared.requestCore(for: bookmark, setup:nil, completionHandler: { (core, _) in
+				if let core = core {
+					core.delegate = self
+
+					func finalize() {
+						OCCoreManager.shared.returnCore(for: bookmark, completionHandler: {
+							self.completed()
+						})
+					}
+
+					core.fetchUpdates(completionHandler: { (_, _) in
+						self.uploadDirectoryTracking = core.trackItem(atPath: path, trackingHandler: { (error, item, _) in
+
+							if error != nil {
+								Log.error("Error \(String(describing: error))")
+							}
+
+							if item != nil {
+								self.uploadMediaAssets(with: core, at: item!, completion: {
+									finalize()
+								})
+							} else {
+								Log.warning("Instant upload directory not found")
+								userDefaults.resetInstantUploadConfiguration()
+								finalize()
+								self.showFeatureDisabledAlert()
+							}
+						})
+					})
+				}
+			})
+		}
+	}
+
+	func core(_ core: OCCore, handleError error: Error?, issue: OCIssue?) {
+		if let error = error {
+			self.result = .failure(error)
+			Log.error("Error \(String(describing: error))")
+		}
+		completed()
+	}
+
+	private func uploadMediaAssets(with core:OCCore, at item:OCItem, completion:@escaping () -> Void) {
+		guard let userDefaults = OCAppIdentity.shared.userDefaults else { return }
+
+		let uploadGroup = DispatchGroup()
+		var assets = [PHAsset]()
+
+		// Add photo assets
+		if let uploadPhotosAfter = userDefaults.instantUploadPhotosAfter {
+			let fetchResult = self.fetchAssetsFromCameraRoll(.images, createdAfter: uploadPhotosAfter)
+			if fetchResult != nil {
+				fetchResult!.enumerateObjects({ (asset, _, _) in
+					assets.append(asset)
+				})
+			}
+		}
+
+		// Add video assets
+		if let uploadVideosAfter = userDefaults.instantUploaVideosAfter {
+			let fetchResult = self.fetchAssetsFromCameraRoll(.videos, createdAfter: uploadVideosAfter)
+			if fetchResult != nil {
+				fetchResult!.enumerateObjects({ (asset, _, _) in
+					assets.append(asset)
+				})
+			}
+		}
+
+		// Perform actual upload operation
+		if assets.count > 0 {
+			uploadGroup.enter()
+			self.upload(assets: assets, with: core, at: item, completion: { () in
+				uploadGroup.leave()
```

Comment:
Not sure if a `DispatchGroup` is needed here. Instead, the `completion`-block here could `OnMainThread`-call this method's `completion`-block directly.

Reply:
@felix-schwarz Indeed, fixed